src/mats_utils/geolocation/coordinates.py

```diff
@@ -178,4 +181,204 @@ def angles(ccditem):
+    t : skyfield.timelib.Time
+        time
+    pos : array[float]
+        position in space where the line starts (~position of MATS). Array of 3 position coordinates in m in the ICRF reference frame
+    FOV : array[float]
+        angle of the line (direction of the line), array of 3 elements in the IFRC reference frame
+            
+    Returns
+    -------
+    res : OptimizeResult object
+        res.x is the distance found in m   
+    """
+    res = minimize_scalar(funheight_square, args=(t, pos, FOV), bracket=(3e5, 8e5))
+    return res
+
+
+def NADIR_geolocation(ccditem,x_step=2,y_step=2):
+    """
+    Function to get the latitude, longitude and solar zenith angle map for each pixel of the image.
+    The values are calculated for some points and then interpolated for each pixel.
+    WARNING : no images are flipped
+    
+    Parameters
+    ----------
+    ccditem : CCDitem
+        measurement
+    x_step : int
+        step along the x-axis in the image between 2 sampled points used for interpolation. The default value is 2.
+    y_step : int
+        step along the y-axis in the image between 2 sampled points used for interpolation. The default value is 2.
+            
+    Returns
+    -------
+    lat_map : array[float]
+        map giving the latitude for each pixel in the image
+    lon_map : array[float]
+        map giving the longitude for each pixel in the image
+    sza_map : array[float]
+        map giving the solar zenith angle for each pixel in the image
+    """
+    im = ccditem['IMAGE']
+    x_deg_map, y_deg_map = deg_map(ccditem) # creating angle deviation map for each pixel (degress)
+    a,b = np.shape(im)
+
+    metoOHB  = R.from_matrix([[0,0,-1],[0,-1,0],[-1,0,0]])
+    ts=sfapi.load.timescale()
+    t=ts.from_datetime(ccditem['EXPDate'].replace(tzinfo=sfapi.utc)) # exposure time  
+    q=ccditem.afsAttitudeState
+    quat=R.from_quat(np.roll(q,-1)) # quaternion of MATS attitude (for the LIMB imager) 
+    pos=ccditem.afsGnssStateJ2000[0:3] # position of MATS
+    
+    xd = range(0,b,x_step) # sampled pixels on the x axis
+    yd = range(0,a,y_step) # sampled pixels on the y axis
+    LAT = np.zeros((len(yd),len(xd)))
+    LON = np.zeros((len(yd),len(xd)))
+    SZA = np.zeros((len(yd),len(xd)))
+
+    # computing the latitude, longitude and solar zenith angles at the intersection of the line of sight and the earth surface
+    # only the line of sights from some sampled pixels are computed
+    for i in range(len(yd)):
+        for j in range(len(xd)):
+                x = xd[j]
+                y = yd[i]
+                # angular transformations
+                # rotation from the line of sight of the LIMB imager to the line of sight of the NADIR pixel
+                angle = R.from_euler('XYZ', [x_deg_map[y,x],-(90-23)+y_deg_map[y,x],0] , degrees=True).apply([1, 0, 0])
+                FOV = quat.apply(metoOHB.apply(angle)) # attitude state for the line of sight of the NADIR pixel    
```

Comment:
Don't know how long this takes  -- it might be possible to vectorise it .I know that R.from_euler kan give multiple rotations , the surface finder could possibly be vectorised or replaced with an analytical expression (not sure if it is worth it yet) and the newp can be a vector so that all the lats and lons come at once,.